executor/adapter.go

```diff
@@ -1337,17 +1337,48 @@ func (a *ExecStmt) GetTextToLog() string {
 	return sql
 }
 
-func (a *ExecStmt) observeStmtBeginForTopSQL() {
+func (a *ExecStmt) observeStmtBeginForTopSQL(ctx context.Context) context.Context {
 	vars := a.Ctx.GetSessionVars()
-	if vars == nil {
-		return
+	sc := vars.StmtCtx
+	normalizedSQL, sqlDigest := sc.SQLDigest()
+	normalizedPlan, planDigest := getPlanDigest(sc, a.Plan)
+	var sqlDigestByte, planDigestByte []byte
+	if sqlDigest != nil {
+		sqlDigestByte = sqlDigest.Bytes()
 	}
-	if stats := a.Ctx.GetStmtStats(); stats != nil && topsqlstate.TopSQLEnabled() {
-		sqlDigest, planDigest := a.getSQLPlanDigest()
-		stats.OnExecutionBegin(sqlDigest, planDigest)
+	if planDigest != nil {
+		planDigestByte = planDigest.Bytes()
+	}
+	stats := a.Ctx.GetStmtStats()
+	if !topsqlstate.TopSQLEnabled() {
+		// Fast plan means the SQL will run very fast, and won't consume too much CPU.
+		// So no need to attach SQL and plan info to catch those running SQL.
+		if isFastPlan(a.Plan) {
+			return ctx
+		}
+		// Always attach the SQL and plan info uses to catch the running SQL when Top SQL is enabled in execution.
+		if stats != nil {
+			sc.KvExecCounter = stats.CreateKvExecCounter(sqlDigestByte, planDigestByte)
+		}
+		return topsql.AttachSQLAndPlanInfo(ctx, sqlDigest, planDigest)
```

Comment:
`OnExecutionBegin` will be skipped executed after this return?

Reply:
Yes, since it has a performance impact when TopSQL is disabled.

Reply:
Looks like there are potential performance improvements here @mornyx @zhongzc

Reply:
In my sysbench point-get test, `OnExecutionBegin` and `OnExecutionFinished` have around %3 QPS impact.

Reply:
Here is the previous implementation:
```go
func (a *ExecStmt) observeStmtBeginForTopSQL() {
	vars := a.Ctx.GetSessionVars()
	if vars == nil {
		return
	}
	if stats := a.Ctx.GetStmtStats(); stats != nil && topsqlstate.TopSQLEnabled() {
		sqlDigest, planDigest := a.getSQLPlanDigest()
		stats.OnExecutionBegin(sqlDigest, planDigest)
		// This is a special logic prepared for TiKV's SQLExecCount.
		vars.StmtCtx.KvExecCounter = stats.CreateKvExecCounter(sqlDigest, planDigest)
	}
}
```
Note `if stats := a.Ctx.GetStmtStats(); stats != nil && topsqlstate.TopSQLEnabled()`. This means that the overhead of `OnExecutionBegin()` does not exist before when TopSQL is not enabled.

Reply:
By the way, when TopSQL is enabled, I think the overhead of `OnExecutionBegin()` is mostly about **lock**. I think we can consider learning `nanolog`'s lock-free-buffer pattern to optimize performance here.

Reply:
How about let's shard the lock into N? It should be very easy to implement and effectively enough to verify the lock cost.

Reply:
> How about let's shard the lock into N? It should be very easy to implement and effectively enough to verify the lock cost.

Can do a test, I think it will work. Since the session is serial, the only scenario where lock contention occurs is the collection of background goroutine. So in theory, segmented locks can effectively reduce the possibility of sessions being blocked by background goroutine.

Reply:
Not from lock. I tried to remove all locks previously and find the performance impact was not reduced much.

Reply:
> Not from lock. I tried to remove all locks previously and found the performance impact was not reduced much.

This seems a bit strange, the logic of `OnExecutionBegin` is very simple (get an item from the map, then increment it). This shouldn't cause a 3% QPS regression...

Reply:
map-assign and map-access have become the main cause. A more complicated lock implementation will introduce extra overhead and worse performance.

Reply:
why would a single map-assign and access cause this kind of drop? It doesn't make sense.